streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -0,0 +1,184 @@
+        this.consumer = consumer;
+        this.time = time;
+        this.partitionValidationTimeoutMs = partitionValidationTimeoutMs;
+    }
+
+
+    @Override
+    public void validatePartitionExists(final TopicPartition topicPartition, final String storeName) {
+        final long start = time.milliseconds();
+        // fetch all info on all topics to avoid multiple remote calls
+        if (partitionInfo.isEmpty()) {
+            try {
+                partitionInfo.putAll(consumer.listTopics());
+            } catch (final TimeoutException e) {
+                log.warn("Could not list topics so will fall back to partition by partition fetching");
+            }
+        }
+
+        final long endTime = time.milliseconds() + partitionValidationTimeoutMs;
+        while (!hasPartition(topicPartition) && time.milliseconds() < endTime) {
+            try {
+                final List<PartitionInfo> partitions = consumer.partitionsFor(topicPartition.topic());
+                if (partitions != null) {
+                    partitionInfo.put(topicPartition.topic(), partitions);
+                }
+            } catch (final TimeoutException e) {
+                throw new StreamsException(String.format("Could not fetch partition info for topic: %s before expiration of the configured request timeout",
+                                                         topicPartition.topic()));
+            }
+        }
+
+        if (!hasPartition(topicPartition)) {
+            throw new StreamsException(String.format("Store %s's change log (%s) does not contain partition %s",
+                                                     storeName, topicPartition.topic(), topicPartition.partition()));
+        }
+        log.debug("Took {} ms to validate that partition {} exists", time.milliseconds() - start, topicPartition);
+    }
+
+    @Override
+    public void register(final StateRestorer restorer) {
+        stateRestorers.put(restorer.partition(), restorer);
+    }
+
+    public void restore() {
+        final long start = time.milliseconds();
+        try {
+            if (!consumer.subscription().isEmpty()) {
+                throw new IllegalStateException(String.format("Restore consumer should have not subscribed to any partitions (%s) beforehand", consumer.subscription()));
+            }
+            final Map<TopicPartition, Long> endOffsets = consumer.endOffsets(stateRestorers.keySet());
+            consumer.assign(stateRestorers.keySet());
+
+            for (final StateRestorer restorer : stateRestorers.values()) {
+                if (restorer.checkpoint() != null) {
+                    consumer.seek(restorer.partition(), restorer.checkpoint());
+                } else {
+                    consumer.seekToBeginning(Collections.singletonList(restorer.partition()));
+                }
+            }
+
+            final Set<TopicPartition> partitions = new HashSet<>(stateRestorers.keySet());
```

Comment:
Is it worthwhile to filter the partitions for those whose offset limit is 0? It will help sending fetch request with less partitions and may result in less `poll` calls.